diff --git a/scope-advice/common.h b/scope-advice/common.h
index 329487d..9eae531 100644
--- a/scope-advice/common.h
+++ b/scope-advice/common.h
@@ -65,12 +65,12 @@
 #endif
 
 // optimization switches
-#define DO_FILTER 1
-#define DO_PARALLEL 1
-#define DO_SAMPLING 1
-#define DO_STREAM 1
+#define DO_FILTER 0
+#define DO_PARALLEL 0
+#define DO_SAMPLING 0
+#define DO_STREAM 0
 // evaluation flag to measure NVBit overhead
-#define DO_ANALYZE 1
+#define DO_ANALYZE 0
 
 /* Below enums are used for the creation of 'info' member in mem_access_t struct.
  * It maintains information necessary for host processing, in a bitwise manner.
